### ISSUE TITLE
[ NEW PAGE ] Transpiler Stage Details / Preset Passmanagers

### ISSUE THREAD

**Issue by javabster** [Qiskit/documentation#104]
## What the new page should cover:

- [ ] Which stages do or do not preserve unitarity equivalence
- [ ] what stuff you need to know from the start: e.g. initial layout before the stage
- [ ] pre/post stages

## SME collaborators (involve for planning and review):
@1ucian0 @haimeng-zhang @mtreinish 

## Existing resources to build upon
- https://qiskit.org/documentation/apidoc/transpiler.html#working-with-preset-pass-managers
- https://qiskit.org/documentation/apidoc/transpiler.html
- https://qiskit.org/documentation/apidoc/transpiler_plugins.html

**Comment by beckykd** [comment]
Here's a discussion from the Defaults topic (https://github.com/Qiskit/documentation/pull/267#discussion_r1379270616) that probably fits   better with the Transpiler stage details section:

(Nick): One thing I think is super confusing is the difference between Layout and Routing (together I call these Mapping, but that might be a personal thing). At a functional level, I think of them as

Routing: insert SWAPs due to the limited connectivity of the lattice
Layout: pick the set of qubits in the lattice which to use based on their noise properties
However, in practice these seem to be somewhat conflated, likely do to needing to pick a set of qubits before you minimize the number of SWAPs? And I think of Sabre as being used for Routing and VF2 as used for Layout. Is this the right place to discuss this, or should be be linking to another doc where this is discussed.

Member
@[mtreinish](https://github.com/mtreinish) mtreinish [19 hours ago](https://github.com/Qiskit/documentation/pull/267#discussion_r1379306207)
Tl;dr It might be worth saving this for some of the details on stages section of the docs, because this makes more sense there talking about how all the stages are constructed. For the definition of what's in the different optimization levels by default even the list of trivial -> vf2 -> sabre might be too much detail for this document.

Yeah, they're tightly coupled. Sabre is actually both a routing algorithm and a layout algorithm (it selects a layout by using a routing algorithm but instead of inserting swap gates it swaps the qubits virtually in the layout). The typical pipeline outlined here for level 1 is we first try the trivial layout (0->0, 1->1, etc) if that matches device connectivity we use it and move on (skipping routing). If it doesn't then we run VF2Layout (which is different from VF2PostLayout) which uses rustworkx to find an isomorphic subgraph on the device to execute the circuit on with no swaps (and if it finds > 1 select the 1 with best noise properties) and also skip routing, then if nothing can be found we run SabreLayout to select a good heuristic layout.

Just to make it extra confusing, SabreLayout also runs routing and inserts swaps. This is done as an efficiency because the whole algorithm is written in rust and it's much faster to compute the layout, find the swaps with that layout, and then select the layout which gives you the best output and just return that as a single thing. Then to make it even more confusing VF2PostLayout runs as part of the routing stage to pick a better layout because it only can be used after routing. But these are implementation details of the specific algorithms we use in the presets and the stages are logically distinct but related at a high level (i.e. you need a layout to run routing because you can't compute where to insert swaps until you know which physical qubits you're running on).

**Comment by beckykd** [comment]
In the Defaults topic, I came across this: init_method ([str](https://docs.python.org/3/library/stdtypes.html#str) | None) – The plugin name to use for the init stage. By default an external plugin is not used. You can see a list of installed plugins by using [list_stage_plugins()](https://qiskit.org/documentation/apidoc/transpiler_plugins.html#qiskit.transpiler.preset_passmanagers.plugin.list_stage_plugins) **with "init" for the stage name argument**.   It's hard to find the list of actual stages (you have to run another API.  Why?  Why not just list the possible values for stage?), but it does seem that `init` is one of them...

which implies that there is an official  "init" or initialization stage.  I see that this topic is going to cover the "pre" stuff, but maybe it should be listed as an actual stage; not just stuff that happens before the first stage.

**Comment by kaelynj** [comment]
> In the Defaults topic, I came across this: init_method ([str](https://docs.python.org/3/library/stdtypes.html#str) | None) – The plugin name to use for the init stage. By default an external plugin is not used. You can see a list of installed plugins by using [list_stage_plugins()](https://qiskit.org/documentation/apidoc/transpiler_plugins.html#qiskit.transpiler.preset_passmanagers.plugin.list_stage_plugins) **with "init" for the stage name argument**. It's hard to find the list of actual stages (you have to run another API. Why? Why not just list the possible values for stage?), but it does seem that `init` is one of them...
> 
> which implies that there is an official "init" or initialization stage. I see that this topic is going to cover the "pre" stuff, but maybe it should be listed as an actual stage; not just stuff that happens before the first stage.

From what I've been able to gather, the `init` stage by default will only unroll 3 qubit gates and do some sanity checks if there are any control flow operations.  It can be customized with a plugin but otherwise doesn't do much by default.  I currently have a draft PR up where I talk about this some more in #283

**Comment by kaelynj** [comment]
Also @nbronn and @mtreinish, let me know what you think about the Layout and Routing stage sections in the pr.  I think I've captured the differences, but need some feedback to be sure.
